ngraph/core/src/op/idft.cpp

```diff
@@ -0,0 +1,196 @@
+// limitations under the License.
+//*****************************************************************************
+
+#include <algorithm>
+#include <memory>
+#include <ngraph/validation_util.hpp>
+
+#include "itt.hpp"
+#include "ngraph/attribute_visitor.hpp"
+#include "ngraph/axis_set.hpp"
+#include "ngraph/axis_vector.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/idft.hpp"
+#include "ngraph/op/util/op_types.hpp"
+#include "ngraph/runtime/host_tensor.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+NGRAPH_RTTI_DEFINITION(op::v7::IDFT, "IDFT", 7);
+
+op::v7::IDFT::IDFT(const Output<Node>& data, const Output<Node>& axes)
+    : Op({data, axes})
+{
+    constructor_validate_and_infer_types();
+}
+
+op::v7::IDFT::IDFT(const Output<Node>& data,
+                   const Output<Node>& axes,
+                   const Output<Node>& signal_size)
+    : Op({data, axes, signal_size})
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::IDFT::visit_attributes(AttributeVisitor& visitor)
+{
+    NGRAPH_OP_SCOPE(v7_IDFT_visit_attributes);
+    return true;
+}
+
+std::shared_ptr<Node> op::v7::IDFT::clone_with_new_inputs(const OutputVector& new_args) const
+{
+    NGRAPH_OP_SCOPE(v7_IDFT_clone_with_new_inputs);
+    check_new_args_count(this, new_args);
+    NODE_VALIDATION_CHECK(
+        this, new_args.size() == 2 || new_args.size() == 3, "Number of inputs must be 2 or 3");
+
+    if (new_args.size() == 2)
+    {
+        return std::make_shared<op::v7::IDFT>(new_args.at(0), new_args.at(1));
+    }
+
+    return std::make_shared<op::v7::IDFT>(new_args.at(0), new_args.at(1), new_args.at(2));
+}
+
+void op::v7::IDFT::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v7_IDFT_validate_and_infer_types);
+
+    size_t num_of_inputs = get_input_size();
+
+    NODE_VALIDATION_CHECK(
+        this, num_of_inputs == 2 || num_of_inputs == 3, "IDFT must have 2 or 3 inputs.");
+
+    element::Type input_et = get_input_element_type(0);
+    NODE_VALIDATION_CHECK(this,
+                          input_et == element::f32 || input_et == element::f16 ||
+                              input_et == element::bf16,
+                          "Input element type must be f32, f16, or bf16");
+
+    element::Type axes_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          axes_et == element::i64 || axes_et == element::i32 ||
+                              axes_et == element::u32 || axes_et == element::u64,
+                          "Axes element type must be i32, i64, u32 or u64");
+
+    PartialShape input_shape = PartialShape(get_input_partial_shape(0));
+    PartialShape axes_shape = PartialShape(get_input_partial_shape(1));
+    PartialShape output_shape = input_shape;
+
+    if (input_shape.rank().is_dynamic() || axes_shape.rank().is_dynamic() ||
+        !is_type<op::Constant>(input_value(1).get_node()))
+    {
+        set_output_type(0, get_input_element_type(0), output_shape);
+        return;
+    }
```

Comment:
If the `signal_size` input is specified (and not all elements are equal to `-1`) then the `output_shape` is not equal to the `input_shape`!  So this check is invalid.

If the `signal_size` input is not specified then we can should `input_shape` as `output_shape`. In the other case we should use values from `axes` input and `signal_size` and calculate output shape.

If the `axes` values are not specified then the output shape tensor should have all elements to be Dynamic, except the last one which should be equal to 2.

Reply:
But we cannot calculate output shape using `signal_size`, if `axes` or `signal_size` are not constants. For eaxmple, `signal_size` is a constant with not all elements are equal to `-1`, but `axes` is not constant.

Reply:
Yes, we cannot calculate them. But we cannot set the output shape to be equal to `input_shape`. The output shape should be {Dynamic, Dynamic,...,Dynamic, 2}.

Reply:
And then, if `axes` is constant and `signal_size` is not constant and has static rank, we should set elements of `output_shape` specified in `axes` as `dynamic()`.

Reply:
Done.